### PR TITLE
New feature: Allows the output data of a layer to remain in the low-level memory for later use when possible

### PR DESCRIPTION
**Pull request by JiacongSun**
### **New feature:**

- This feature allows the output data of a layer to remain in the low-level memory when possible. This enables direct use of the output data by the next layer, eliminating the need for data to travel to/from the top memory level.

**Note:** 
1. This feature applies only to layers on the same branch.
2. By default, it is assumed that the initial input and final output of the entire network can be generated from the low-level memory. If your case requires them to travel to/from the top memory level, you can change this assumption by setting the `workload_data_always_from_top_mem` parameter to `True` in the `run()` function within the `SearchNoUseMemStage.py` file.

### **New stages:**
- SearchNoUseMemStage: This stage searches for unnecessary top memory levels for each layer and generates a pointer indicating which level the output data of each layer should travel up to.
- RemoveNoUseMemStage: This stage removes memory instances with a level higher than the pointer.

### **How to use:**
- Place the **SearchNoUseMemStage** before the **WorkloadStage**, and place the **RemoveNoUseMemStage** after the **WorkloadStage**.

### **Example:**
An example function, `get_hardware_performance_zigzag_unused_mem_removing` is provided in `api.py` for reference.

### CODE REVIEW

File: zigzag/classes/stages/RemoveNoUseMemStage.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+##     remove ['O'] in mem.served_operand, mem_port_alloc
+##   if ['W'] in mem.served_operand and mem.mem_level > target_const_mem_level:
+##     remove ['W'] in mem.served_operand, mem_port_alloc
+## 2. Remove no-use memory
+## for mem in mem_levels(sort_order: from top to bottom):
+##   if mem.served_operand == empty:
+##     do not add the current mem into the modified architecture
+#####################################################
+
+class RemoveNoUseMemStage(Stage):
```
Comment (asyms) [inline zigzag/classes/stages/RemoveNoUseMemStage.py]: This name is not the most informative. Maybe something like "RemoveUnusedMemoryStage"?

File: zigzag/classes/stages/RemoveNoUseMemStage.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    def run(self) -> Generator:
+        modified_accelerator = self.generate_accelerator_removing_nouse_mem()
+        sub_stage = self.list_of_callables[0](self.list_of_callables[1:],
+                accelerator=modified_accelerator,
+                layer=self.layer,
+                **self.kwargs,)
+        for cme, extra_info in sub_stage.run():
+            yield cme, extra_info
+
+    def generate_accelerator_removing_nouse_mem(self):
```
Comment (asyms) [inline zigzag/classes/stages/RemoveNoUseMemStage.py]: Similarly rename this function.

File: zigzag/classes/stages/RemoveNoUseMemStage.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        ## Remove nouse memory level according to update_mem_list and mem_update_weight
+        curr_id = self.layer_list[self.layer]  # current layer id (key) in mem_udpate_list
+        output_operand = self.layer.memory_operand_links[self.layer.output_operand]  # output representation
+        core = next(iter(self.accelerator.cores))
+        operational_array = core.operational_array
+        memory_hierarchy = core.memory_hierarchy
+
+        if len(self.layer.constant_operands) == 1:
+            act_operand = self.layer.memory_operand_links[[operand for operand in self.layer.input_operands if operand not in self.layer.constant_operands][0]]  # act representation
```
Comment (asyms) [inline zigzag/classes/stages/RemoveNoUseMemStage.py]: Rename all operand variables to reflect if they're "layer operands" or "memory operands" either in the name or in comments

File: zigzag/classes/stages/SearchNoUseMemStage.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+##         if sum(layer[operand_size] for operand in mem.operands) <= mem.size:
+##           if ["I", "O"] both in mem.operands:
+##             mem_update_list[layer]["O"] = current_mem_level
+##             if layer.index == 0: ## the 1st execution layer
+##               mem_update_list[layer]["I"] = current_mem_level
+##           if ("W" in mem.operand) and (current_mem_level < mem_update_weight):
+##             mem_update_weight = current_mem_level
+#####################################################
+
+class SearchNoUseMemStage(Stage):
```
Comment (asyms) [inline zigzag/classes/stages/SearchNoUseMemStage.py]: Rename this class and documentation to "SearchUnusedMemoryStage" or something similar.

File: zigzag/classes/stages/RemoveNoUseMemStage.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+                if pr_loop_keys[0] in related_loop:
+                    act_operand = operand
+            weight_operand: list = [x for x in self.layer.constant_operands if x != act_operand]
+            assert len(weight_operand) == 1
+            weight_operand: str = weight_operand[0]
+            act_operand = self.layer.memory_operand_links[act_operand]  # map from layer representation to hardware representation
+            const_operand = self.layer.memory_operand_links[weight_operand]  # weight representation
+
+        # Find target_act/const/output_mem_level
+        for pos, ele in enumerate(self.mem_update_list[f"{curr_id}"]):
```
Comment (asyms) [inline zigzag/classes/stages/RemoveNoUseMemStage.py]: is curr_id not already a string? If not, you should convert a variable to a string using str(...)

File: zigzag/classes/stages/RemoveNoUseMemStage.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            if list(ele.keys())[0] == f"{act_operand}":
+                target_act_mem_level = self.mem_update_list[f"{curr_id}"][pos][f"{act_operand}"]
```
Comment (asyms) [inline zigzag/classes/stages/RemoveNoUseMemStage.py]: Pretty sure act_operand is already a string?

File: zigzag/classes/stages/RemoveNoUseMemStage.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                    port_alloc=new_port_alloc,
+                    served_dimensions=new_served_dimensions,
+                )
+
+        # Create the new core
+        id = core.id
+        dataflows = core.dataflows
+        if dataflows is not None:
+            raise NotImplementedError(
+                "Scale your core-defined dataflows accordingly here."
```
Comment (asyms) [inline zigzag/classes/stages/RemoveNoUseMemStage.py]: Why do the dataflows have to be scaled?

File: zigzag/classes/stages/SearchNoUseMemStage.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+                **self.kwargs,)
+        for cme, (layer, extra_info) in sub_stage.run():
+            yield cme, (layer, extra_info)
+
+    def update_top_mem_level(self):
+        """
+        Update mem_update_list and mem_update_weight according to the algorithm description at the file beginning.
+        """
+        """
+        param const_operand: constant operand name (e.g. "W")
```
Comment (asyms) [inline zigzag/classes/stages/SearchNoUseMemStage.py]: You specify three parameters but the function has no parameters?

**Review by asyms — CHANGES_REQUESTED** [review]
Please take a look at the comments, and format your code using [black](https://pypi.org/project/black/)

**Comment by JiacongSun** [comment]
**New fixes:**

1. SearchNoUseMemStage is renamed as SearchUnusedMemoryStage, RemoveNoUseMemStage is renamed as RemoveUnusedMemoryStage.
2. Coding style in stages above are reformated using black.
3. Add more comments in the new stages, including explicitly label whether a operand is of the layer representation or the mem representation.
4. pytest cases are added under `/tests/main/test_without_unused_memory/`. Original pytest cases are moved to `/tests/main/test_origin/`.

**Extra note:**

- Only when user-defined workload (.py) is used **and** there are Adder layers within the workload, the estimation of SearchUnusedMemoryStage for these Adder layers is more pessimistic, which means the topest memory level for output of these Adder layers possibly is higher than expected. (For details, please refer to the comments in SearchUnusedMemoryStage).
- For other cases (.onnx workload or non-Adder layers), there is no problem.

**Review by asyms — APPROVED** [review]
Looks good to me!

**Pull request merged**